go/vt/vtgate/planbuilder/jointree.go

```diff
@@ -633,3 +643,48 @@ func (jp *joinPlan) pushOutputColumns(columns []*sqlparser.ColName, semTable *se
 	}
 	return outputColumns
 }
+
+// costFor returns a cost struct to make route choices easier to compare
+func costFor(foundVindex vindexes.Vindex, opcode engine.RouteOpcode) cost {
+	switch opcode {
+	// For these opcodes, we should not have a vindex, so we just return the opcode as the cost
+	case engine.SelectUnsharded, engine.SelectNext, engine.SelectDBA, engine.SelectReference, engine.SelectNone, engine.SelectScatter:
+		return cost{
+			opCode: opcode,
+		}
+	}
+
+	// if we have a multiplier that is non-zero, we should have a vindex
+	if foundVindex == nil {
+		panic("expected a vindex")
+	}
+
+	return cost{
+		vindexCost: foundVindex.Cost(),
+		isUnique:   foundVindex.IsUnique(),
+		opCode:     opcode,
+	}
+}
+
+// opCodePrecedence tracks the order that op codes are compared to each other
+var opCodePrecedence = map[engine.RouteOpcode]int{
+	engine.SelectEqualUnique: 1,
+	engine.SelectEqual:       2,
+	engine.SelectIN:          3,
+	engine.SelectMultiEqual:  4,
+	engine.SelectScatter:     5,
+}
+
+// less compares two costs and returns true if the first cost is cheaper than the second
+func less(c1, c2 cost) bool {
+	precedence1 := opCodePrecedence[c1.opCode]
+	precedence2 := opCodePrecedence[c2.opCode]
+	switch {
+	case precedence1 != precedence2:
+		return precedence1 < precedence2
+	case c1.isUnique == c2.isUnique:
+		return c1.vindexCost <= c2.vindexCost
+	default:
+		return c1.isUnique
+	}
+}
```

Comment:
I think this whole chunk can be simplified. The `engine.RouteOpcode` is an enumeration and follows the same increment as `opCodePrecedence`. The plan tests were still passing after using directly `c1.opCode` and `c2.opCode`, instead of `precedence1` and `precedence2`.


```suggestion
// less compares two costs and returns true if the first cost is cheaper than the second
func less(c1, c2 cost) bool {
	switch {
	case c1.opCode != c2.opCode:
		return c1.opCode < c2.opCode
	case c1.isUnique == c2.isUnique:
		return c1.vindexCost <= c2.vindexCost
	default:
		return c1.isUnique
	}
}
```

Reply:
hmm... that is not a bad idea. :+1: